File: src/test/java/no/nav/bidrag/commons/web/CorrelationIdFilterTest.java
package no.nav.bidrag.commons.web;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.lenient;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.Appender;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponseWrapper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatcher;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@DisplayName("CorrelationIdFilter")
@ExtendWith(MockitoExtension.class)
class CorrelationIdFilterTest {

  private static final String CORRELATION_ID = "X-Correlation-ID";

  private final CorrelationIdFilter correlationIdFilter = new CorrelationIdFilter();
  private final Set<String> logMeldinger = new HashSet<>();

  @Mock
  @SuppressWarnings("rawtypes")
  private Appender appenderMock;

  @Mock
  private FilterChain filterChainMock;

  @Mock
  private HttpServletRequest httpServletRequestMock;

  @Mock
  private HttpServletResponseWrapper httpServletResponseMock;

  @BeforeEach
  @SuppressWarnings("unchecked")
  void mockLogAppender() {
    ch.qos.logback.classic.Logger logger = (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
    lenient().when(appenderMock.getName()).thenReturn("MOCK");
    lenient().when(appenderMock.isStarted()).thenReturn(true);
    logger.addAppender(appenderMock);
  }

  @Test
  @SuppressWarnings({"unchecked", "rawtypes"})
  @DisplayName("skal logge requests mot servlet")
  void skalLoggeRequestsMotActuatorEndpoints() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("something");
    when(httpServletRequestMock.getMethod()).thenReturn("GET");

    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    assertAll(
        () -> verify(appenderMock, atLeastOnce()).doAppend(
            argThat((ArgumentMatcher) argument -> {
              logMeldinger.add(((ILoggingEvent) argument).getFormattedMessage());

              return true;
            })),
        () -> assertThat(String.join("\n", logMeldinger)).contains("prosessing GET something")
    );
  }

  @Test
  @DisplayName("skal ikke logge requests mot actuator endpoints")
  void skalIkkeLoggeRequestsMotSwaggerResources() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/actuator/health").thenReturn("/actuator/something");

    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    verifyNoInteractions(appenderMock);
  }

  @Test
  @DisplayName("skal ikke logge requests etter api-docs")
  void skalIkkeLoggeRequestsEtterApiDocs() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/v3/api-docs/").thenReturn("/v2/api-docs/");

    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    verifyNoInteractions(appenderMock);
  }

  @Test
  @DisplayName("skal ikke logge requests mot swagger resources")
  void skalIkkeLoggeRequestsMotActuatorEndpoints() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/swagger-ui-bundle.js").thenReturn("/swagger-config");

    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    verifyNoInteractions(appenderMock);
  }

  @Test
  @DisplayName("skal legge HttpHeader.CORRELATION_ID på response")
  void skalLeggeHttpHeaderCorrelationIdPaaResponse() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("somewhere");

    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    verify(httpServletResponseMock).addHeader(eq(CORRELATION_ID), anyString());
  }

  @Test
  @DisplayName("skal ikke legge HttpHeader.CORRELATION_ID på response når den allerede eksisterer")
  void skalIkkeLeggeHttpHeaderCorrelationIdPaaResponseNaarDenAlleredeEksisterer() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("somewhere else");
    when(httpServletRequestMock.getHeader(CORRELATION_ID)).thenReturn("svada");

    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    verify(httpServletResponseMock, never()).addHeader(anyString(), anyString());
  }

  @Test
  @DisplayName("skal bruke request uri som correlation id")
  void skalBrukeRequestUriSomCorrelationId() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("somewhere");
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    ArgumentCaptor<String> correlationCaptor = ArgumentCaptor.forClass(String.class);
    verify(httpServletResponseMock).addHeader(eq(CORRELATION_ID), correlationCaptor.capture());

    assertThat(correlationCaptor.getValue()).contains("-somewhere");
  }

  @Test
  @DisplayName("skal lage correlation id av siste del fra request uri")
  void skalLageCorrelationIdAvSisteDelFraRequestUri() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/en/forbanna/journalpost");
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    ArgumentCaptor<String> correlationCaptor = ArgumentCaptor.forClass(String.class);
    verify(httpServletResponseMock).addHeader(eq(CORRELATION_ID), correlationCaptor.capture());

    assertThat(correlationCaptor.getValue()).contains("-journalpost");
  }

  @Test
  @DisplayName("skal lage correlation id av siste del fra request uri som er ren tekst, samt legge på eventuelle tall")
  void skalLageCorrelationIdAvSisteDelFraRequestUriSomHarRenTekstSamtEventuelleTall() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/en/identifisert/journalpost/1001");
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    ArgumentCaptor<String> correlationCaptor = ArgumentCaptor.forClass(String.class);
    verify(httpServletResponseMock).addHeader(eq(CORRELATION_ID), correlationCaptor.capture());

    assertThat(correlationCaptor.getValue()).contains("-journalpost/1001");
  }

  @Test
  @DisplayName("skal lage correlation id av siste del fra request uri som er ren tekst, samt legge på eventuelle prefiksede tall")
  void skalLageCorrelationIdAvSisteDelFraRequestUriSomHarRenTekstSamtEventuellePrefiksedeTall() throws IOException, ServletException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("/en/identifisert/journalpost/BID-1001");
    correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock);

    ArgumentCaptor<String> correlationCaptor = ArgumentCaptor.forClass(String.class);
    verify(httpServletResponseMock).addHeader(eq(CORRELATION_ID), correlationCaptor.capture());

    assertThat(correlationCaptor.getValue()).contains("-journalpost/BID-1001");
  }

  @Test
  @DisplayName("skal legge correlation id på ThreadLocal som kan leses for konfigurasjon")
  void skalLeggeCorrelationIdPaaThreadLocal() throws InterruptedException {
    when(httpServletRequestMock.getRequestURI()).thenReturn("go somewhere");

    CorrelationIdThread aCorrelationIdThread = new CorrelationIdThread(
        () -> correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
    );

    CorrelationIdThread anotherCorrelationIdThread = new CorrelationIdThread(
        () -> correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
    );

    aCorrelationIdThread.start();
    Thread.sleep(10); // to be sure the value is not from the same millis
    anotherCorrelationIdThread.start();
    aCorrelationIdThread.join();
    anotherCorrelationIdThread.join();

    assertAll(
        () -> assertThat(aCorrelationIdThread.correlationId).as("a correlation id")
            .isNotNull().isNotEqualTo(anotherCorrelationIdThread.correlationId),
        () -> assertThat(anotherCorrelationIdThread.correlationId).as("another correlation id")
            .isNotNull().isNotEqualTo(aCorrelationIdThread.correlationId)
    );
  }

  static class CorrelationIdThread extends Thread {

    private final FilterExecutor filterExecutor;
    String correlationId;

    CorrelationIdThread(FilterExecutor filterExecutor) {
      this.filterExecutor = filterExecutor;
    }

    @Override
    public void run() {
      try {
        filterExecutor.doFilter();
        correlationId = CorrelationIdFilter.fetchCorrelationIdForThread();
      } catch (IOException | ServletException e) {
        throw new AssertionError(e);
      }
    }
  }

  @FunctionalInterface
  interface FilterExecutor {

    void doFilter() throws IOException, SecurityException, ServletException;
  }
}

File: src/test/kotlin/no/nav/bidrag/commons/web/CorrelationIdFilterTest.kt
package no.nav.bidrag.commons.web

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.core.Appender
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponseWrapper
import no.nav.bidrag.commons.web.CorrelationIdFilter.Companion.fetchCorrelationIdForThread
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.LoggerFactory
import java.io.IOException

internal class CorrelationIdFilterTest {
    private val correlationIdFilter = CorrelationIdFilter()
    private val logMeldinger: MutableList<ILoggingEvent> = ArrayList()

    private val appenderMock: Appender<ILoggingEvent> = mockk(relaxed = true)

    private val filterChainMock: FilterChain = mockk(relaxed = true)

    private val httpServletRequestMock: HttpServletRequest = mockk(relaxed = true)

    private val httpServletResponseMock: HttpServletResponseWrapper = mockk(relaxed = true)

    @BeforeEach
    fun mockLogAppender() {
        val logger = LoggerFactory.getLogger(CorrelationIdFilter::class.java) as Logger
        logger.level = Level.DEBUG
        every { appenderMock.name } returns "MOCK"
        every { appenderMock.isStarted } returns true
        logger.addAppender(appenderMock)
    }

    @Test
    fun `skal logge requests mot servlet`() {
        every { httpServletRequestMock.requestURI } returns "something"
        every { httpServletRequestMock.method } returns "GET"
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        verify(atLeast = 1) { appenderMock.doAppend(capture(logMeldinger)) }
        logMeldinger.joinToString { it.formattedMessage } shouldContain "prosessing GET something"
    }

    @Test
    fun `skal ikke logge requests mot actuator endpoints`() {
        every { httpServletRequestMock.requestURI } returns "/actuator/health" andThen "/actuator/something"
        every { httpServletRequestMock.method } returns "GET"
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        verify(exactly = 0) { appenderMock.doAppend(any()) }
    }

    @Test
    fun `skal ikke logge requests etter api-docs`() {
        every { httpServletRequestMock.requestURI } returns "/v3/api-docs/" andThen "/v2/api-docs/"
        every { httpServletRequestMock.method } returns "GET"
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        verify(exactly = 0) { appenderMock.doAppend(any()) }
    }

    @Test
    fun `skal ikke logge requests mot swagger resources`() {
        every { httpServletRequestMock.requestURI } returns "/swagger-ui-bundle.js" andThen "/swagger-config"
        every { httpServletRequestMock.method } returns "GET"
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        verify(exactly = 0) { appenderMock.doAppend(any()) }
    }

    @Test
    fun `skal legge HttpHeader CORRELATION_ID på response`() {
        every { httpServletRequestMock.requestURI } returns "somewhere"
        every { httpServletRequestMock.method } returns "GET"
        every { httpServletRequestMock.getHeader(CorrelationIdFilter.CORRELATION_ID_HEADER) } returns null
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        verify { httpServletResponseMock.addHeader(CORRELATION_ID, any()) }
    }

    @Test
    fun `skal ikke legge HttpHeader CORRELATION_ID på response når den allerede eksisterer`() {
        every { httpServletRequestMock.requestURI } returns "somewhere else"
        every { httpServletRequestMock.method } returns "GET"
        every { httpServletRequestMock.getHeader(CORRELATION_ID) } returns "svada"
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        verify(exactly = 0) { httpServletResponseMock.addHeader(any(), any()) }
    }

    @Test
    fun `skal bruke request uri som correlation id`() {
        every { httpServletRequestMock.requestURI } returns "somewhere"
        every { httpServletRequestMock.method } returns "GET"
        every { httpServletRequestMock.getHeader(CorrelationIdFilter.CORRELATION_ID_HEADER) } returns null
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        val correlationCaptor = slot<String>()
        verify { httpServletResponseMock.addHeader(CORRELATION_ID, capture(correlationCaptor)) }
        correlationCaptor.captured shouldContain "-somewhere"
    }

    @Test
    fun `skal lage correlation id av siste del fra request uri`() {
        every { httpServletRequestMock.requestURI } returns "/en/forbanna/journalpost"
        every { httpServletRequestMock.method } returns "GET"
        every { httpServletRequestMock.getHeader(CorrelationIdFilter.CORRELATION_ID_HEADER) } returns null
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        val correlationCaptor = slot<String>()
        verify { httpServletResponseMock.addHeader(CORRELATION_ID, capture(correlationCaptor)) }
        correlationCaptor.captured shouldContain "-journalpost"
    }

    @Test
    fun `skal lage correlation id av siste del fra request uri som er ren tekst, samt legge på eventuelle tall`() {
        every { httpServletRequestMock.requestURI } returns "/en/identifisert/journalpost/1001"
        every { httpServletRequestMock.getHeader(CorrelationIdFilter.CORRELATION_ID_HEADER) } returns null
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        val correlationCaptor = slot<String>()
        verify { httpServletResponseMock.addHeader(CORRELATION_ID, capture(correlationCaptor)) }
        correlationCaptor.captured shouldContain "-journalpost/1001"
    }

    @Test
    fun `skal lage correlation id av siste del fra request uri som er ren tekst, samt legge på eventuelle prefiksede tall`() {
        every { httpServletRequestMock.requestURI } returns "/en/identifisert/journalpost/BID-1001"
        every { httpServletRequestMock.method } returns "GET"
        every { httpServletRequestMock.getHeader(CorrelationIdFilter.CORRELATION_ID_HEADER) } returns null
        correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock)
        val correlationCaptor = slot<String>()
        verify { httpServletResponseMock.addHeader(CORRELATION_ID, capture(correlationCaptor)) }
        correlationCaptor.captured shouldContain "-journalpost/BID-1001"
    }

    @Test
    fun `skal legge correlation id på ThreadLocal som kan leses for konfigurasjon`() {
        every { httpServletRequestMock.requestURI } returns "go somewhere"
        val aCorrelationIdThread =
            CorrelationIdThread { correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock) }
        val anotherCorrelationIdThread =
            CorrelationIdThread { correlationIdFilter.doFilter(httpServletRequestMock, httpServletResponseMock, filterChainMock) }
        aCorrelationIdThread.start()
        Thread.sleep(10) // to be sure the value is not from the same millis
        anotherCorrelationIdThread.start()
        aCorrelationIdThread.join()
        anotherCorrelationIdThread.join()
        aCorrelationIdThread.correlationId shouldNotBe null shouldNotBe anotherCorrelationIdThread.correlationId
        anotherCorrelationIdThread.correlationId shouldNotBe null shouldNotBe aCorrelationIdThread.correlationId
    }

    internal class CorrelationIdThread(private val filterExecutor: FilterExecutor) : Thread() {
        var correlationId: String? = null
        override fun run() {
            correlationId = try {
                filterExecutor.doFilter()
                fetchCorrelationIdForThread()
            } catch (e: IOException) {
                throw AssertionError(e)
            } catch (e: ServletException) {
                throw AssertionError(e)
            }
        }
    }

    internal fun interface FilterExecutor {

        fun doFilter()
    }

    companion object {
        private const val CORRELATION_ID = "X-Correlation-ID"
    }
}
